tests/chainer_tests/functions_tests/connection_tests/test_linear.py

```diff
@@ -41,138 +33,70 @@ def _to_noncontiguous(arrays):
+        # TODO(crcrpar): Remove this relaxation when
+        # a known issue in the reduction of ChainerX is resolved.
+        if test_name == 'test_forward':
+            if (self.x_dtype == numpy.float16 and
+                    self.W_dtype == numpy.float16 and
+                    self.n_batch_axes == 3 and
+                    self.backend_config.use_chainerx and
+                    self.backend_config.chainerx_device == 'native:0'):
+                self.check_forward_options['atol'] = 5e-3
+
+    def generate_inputs(self):
         data_shape = self.x_shape['data_shape']
+        batch_shape = (4,) + (2,) * (self.n_batch_axes - 1)
+        x = numpy.random.uniform(
+            -1, 1, batch_shape + data_shape).astype(self.x_dtype)
         input_size = numpy.prod(data_shape)
         W = numpy.random.uniform(-1, 1, (2, input_size)).astype(self.W_dtype)
         if self.nobias:
-            b = None
+            return x, W
         else:
             b = numpy.random.uniform(-1, 1, 2).astype(self.x_dtype)
+            return x, W, b
 
-        batch_shape = (4,) + (2,) * (self.n_batch_axes - 1)
-        x = numpy.random.uniform(
-            -1, 1, batch_shape + data_shape).astype(self.x_dtype)
-        gy = numpy.random.uniform(
-            -1, 1, batch_shape + (2,)).astype(self.x_dtype)
-        ggx = numpy.random.uniform(-1, 1, x.shape).astype(self.x_dtype)
-        ggW = numpy.random.uniform(-1, 1, W.shape).astype(self.W_dtype)
-        if self.nobias:
-            ggb = None
-        else:
-            ggb = numpy.random.uniform(-1, 1, b.shape).astype(self.x_dtype)
-        self.inputs = [x, W, b]
-        self.grad_outputs = [gy]
-        self.grad_grad_inputs = [ggx, ggW, ggb]
-        if self.x_dtype == numpy.float16:
-            self.check_forward_options = {'atol': 1e-3, 'rtol': 1e-2}
-            self.check_backward_options = {
-                'dtype': numpy.float64, 'atol': 1e-2, 'rtol': 1e-2}
-            self.check_double_backward_options = {
-                'dtype': numpy.float64, 'atol': 1e-2, 'rtol': 1e-2}
-        elif self.W_dtype == numpy.float16:
-            self.check_forward_options = {}
-            self.check_backward_options = {
-                'dtype': numpy.float64, 'atol': 1e-2, 'rtol': 1e-2}
-            self.check_double_backward_options = {
-                'dtype': numpy.float64, 'atol': 1e-2, 'rtol': 1e-2}
-        else:
-            self.check_forward_options = {}
-            self.check_backward_options = {
-                'atol': 1e-2, 'rtol': 1e-2}
-            self.check_double_backward_options = {
-                'atol': 1e-2, 'rtol': 1e-2}
-
-    def forward_cpu(self, inputs):
-        x, W, b = inputs
+    def generate_grad_outputs(self, outputs_template):
+        dtypes = [self.x_dtype, self.W_dtype, self.x_dtype]
+        return tuple([
+            numpy.random.uniform(-1, 1, a.shape).astype(dtype)
+            for a, dtype in zip(outputs_template, dtypes)
```

Comment:
This method is supposed to return upstream gradients for the output of the function. I think there's only one output from linear but are you supposed to zip with all dtypes of the inputs?

Reply:
You are right and this method must be removed.

background: I might have implemented this in order to clarify the dtypes of (x, W, b) because I have been confused with the production of x_dtype and W_dtype and I have not understood dtype of bias term.

Reply:
Great. Is the background that you mention resolved? Regarding the bias term, it's created with the same dtype as x?

Reply:
Yes, bias term should have the same dtype as x.